rust/src/model.rs

```diff
@@ -0,0 +1,569 @@
+use std::convert::TryFrom;
+
+use num::{bigint::BigInt, clamp, rational::Ratio, traits::identities::Zero};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use crate::{
+    mask::{config::MaskConfig, seed::MaskSeed, Integers, MaskedModel},
+    utils::generate_integer,
+    PetError,
+};
+
+/// Masking of models.
+pub trait MaskModels<N> {
+    /// Get a reference to the weights.
+    fn weights(&self) -> &Vec<N>;
+
+    /// Cast the weights as ratios. Must handle non-finite weights.
+    fn as_ratios(&self) -> Vec<Ratio<BigInt>>;
+
+    /// Mask the model wrt the mask configuration. Enforces bounds on the scalar and weights.
+    fn mask(&self, scalar: f64, config: &MaskConfig) -> (MaskSeed, MaskedModel) {
+        let scalar = &Ratio::<BigInt>::from_float(clamp(scalar, 0_f64, 1_f64)).unwrap();
+        let negative_bound = &-config.add_shift();
+        let positive_bound = config.add_shift();
+        let mask_seed = MaskSeed::generate();
+        let mut prng = ChaCha20Rng::from_seed(mask_seed.as_array());
+        let masked_weights = self
+            .as_ratios()
+            .iter()
+            .map(|weight| {
+                let scaled = scalar * clamp(weight, negative_bound, positive_bound);
+                let shifted = ((scaled + config.add_shift()) * config.exp_shift())
+                    .to_integer()
+                    .to_biguint()
+                    // safe unwrap: shifted weight is guaranteed to be non-negative
+                    .unwrap();
+                (shifted + generate_integer(&mut prng, config.order())) % config.order()
+            })
+            .collect();
+        // safe unwrap: masked weights are guaranteed to conform to the mask configuration
+        let masked_model = MaskedModel::from_parts(masked_weights, config.clone()).unwrap();
+        (mask_seed, masked_model)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
```

Comment:
We usually derive `Eq` along with `PartialEq`. As long as the numbers are not float that should work (equality is not reflexive for floats).

Reply:
i'll add that.